channeldb/kvdb/etcd/commit_queue.go

```diff
@@ -83,27 +85,38 @@ func (c *commitQueue) Add(commitLoop func(), rset readSet, wset writeSet) {
 	}
 
 	if blocked {
-		// Add the transaction to the queue if conflicts with an already
-		// queued one.
+		// Add the transaction to the queue if it conflicts with an
+		// already queued one. It is safe to do so outside the lock,
+		// since this we know it will be executed serially.
 		c.mx.Unlock()
 
-		select {
-		case c.queue <- commitLoop:
-		case <-c.ctx.Done():
-		}
+		c.queueCond.L.Lock()
+		c.queue.PushBack(commitLoop)
+		c.queueCond.L.Unlock()
 	} else {
 		// To make sure we don't add a new tx to the queue that depends
-		// on this "unblocked" tx, grab the commitMutex before lifting
-		// the mutex guarding the lock maps.
-		c.commitMutex.RLock()
+		// on this "unblocked" tx. Increment our free counter before
+		// unlocking so that the mainLoop stops pulling off blocked
+		// transactions from the queue.
+
+		c.queueCond.L.Lock()
+		c.freeCount++
+		c.queueCond.L.Unlock()
+
 		c.mx.Unlock()
 
-		// At this point we're safe to execute the "unblocked" tx, as
-		// we cannot execute blocked tx that may have been read from the
-		// queue until the commitMutex is held.
-		commitLoop()
+		// At this point it is safe to execute the "unblocked" tx, as no
+		// blocked tx will be read from the queue until the freeCount is
+		// decremented back to 0.
+		go func() {
```

Comment:
I'm not sure we need this goroutine normally? It's handy for the commit queue unit test though. Did a few benchmarks with/without and performance difference is negligible.